services/director-v2/src/simcore_service_director_v2/modules/db/repositories/comp_runs.py

```diff
@@ -76,39 +77,44 @@ async def create(
         user_id: UserID,
         project_id: ProjectID,
         cluster_id: ClusterID,
-        iteration: Optional[PositiveInt] = None,
+        iteration: PositiveInt | None = None,
     ) -> CompRunsAtDB:
-        async with self.db_engine.acquire() as conn:
-            if iteration is None:
-                # let's get the latest if it exists
-                last_iteration = await conn.scalar(
-                    sa.select([comp_runs.c.iteration])
-                    .where(
-                        (comp_runs.c.user_id == user_id)
-                        & (comp_runs.c.project_uuid == str(project_id))
+        try:
+            async with self.db_engine.acquire() as conn:
+                if iteration is None:
+                    # let's get the latest if it exists
+                    last_iteration = await conn.scalar(
+                        sa.select([comp_runs.c.iteration])
+                        .where(
+                            (comp_runs.c.user_id == user_id)
+                            & (comp_runs.c.project_uuid == str(project_id))
+                        )
+                        .order_by(desc(comp_runs.c.iteration))
                     )
-                    .order_by(desc(comp_runs.c.iteration))
-                )
-                iteration = (last_iteration or 0) + 1
+                    iteration = (last_iteration or 0) + 1
 
-            result = await conn.execute(
-                comp_runs.insert()  # pylint: disable=no-value-for-parameter
-                .values(
-                    user_id=user_id,
-                    project_uuid=f"{project_id}",
-                    cluster_id=cluster_id if cluster_id != DEFAULT_CLUSTER_ID else None,
-                    iteration=iteration,
-                    result=RUNNING_STATE_TO_DB[RunningState.PUBLISHED],
-                    started=datetime.utcnow(),
+                result = await conn.execute(
+                    comp_runs.insert()  # pylint: disable=no-value-for-parameter
+                    .values(
+                        user_id=user_id,
+                        project_uuid=f"{project_id}",
+                        cluster_id=cluster_id
+                        if cluster_id != DEFAULT_CLUSTER_ID
+                        else None,
+                        iteration=iteration,
+                        result=RUNNING_STATE_TO_DB[RunningState.PUBLISHED],
+                        started=datetime.utcnow(),
+                    )
+                    .returning(literal_column("*"))
                 )
-                .returning(literal_column("*"))
-            )
-            row = await result.first()
-            return CompRunsAtDB.from_orm(row)
+                row = await result.first()
+                return CompRunsAtDB.from_orm(row)
+        except ForeignKeyViolation as exc:
+            raise ClusterNotFoundError(cluster_id=cluster_id) from exc
```

Comment:
THOUGHT: When can this happen? When a cluster is referenced and still not created, right? How do you know this is not from another key violation? I understand that by the context, here is the post probable cause but it would be nice to start developing a relying mechanism to handle pg errors and this could be the perfect excuse :-)

Suggestion: Note that all psycopg errors have some members: ``pgerror``, ``pgcode``, ``cursor`` and ``diag``
https://www.psycopg.org/docs/module.html#psycopg2.Error

Check in particular ``diag`` which is a ``Diagnostic`` object. There we might find that the constraint that failed was e.g. ``fk_cluster_to_groups_gid_groups``

Reply:
use-case is: use the osparc API with the wrong cluster ID